release-25.2: optbuilder: disable mutation of vector indexes while backfilling

mw5h opened the pull request:
Backport 1/1 commits from #144520.

/cc @cockroachdb/release

---

Merging concurrent writes on vector indexes while backfill is in progress is somewhat more complicated problem than can be solved in the 25.2 timeframe. In order to ensure data integrity, we will disable mutation on vector indexes while backfill is in progress.

Informs: #144443
Release note: None

Release justification: This is a safety feature to prevent creating corrupted vector indexes.

mw5h committed:
optbuilder: disable mutation of vector indexes while backfilling

Merging concurrent writes on vector indexes while backfill is in
progress is somewhat more complicated problem than can be solved in the
25.2 timeframe. In order to ensure data integrity, we will disable
mutation on vector indexes while backfill is in progress.

Informs: #144443
Release note (sql change): Vector indexes in 25.2 do not support
mutation while being created with CREATE INDEX or rebuilt with ALTER
PRIMARY KEY. To prevent inadvertant application downtime, both of
these commands require that sql_safe_updates be set to false while
operating on a vector index.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144601)
<!-- Reviewable:end -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for adding this. I think it needs a release note, and I had one suggestion which you can take or leave, but I think it's good.

> Reviewed 13 of 13 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144601#-:-OO13tLECfbSBu9atXmV:b-852p4d)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`-- commits` line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144601#-OO144jvCkKN1iAOgvFT:-OO144jvCkKN1iAOgvFU:b-4y0bin):*
We probably want a release note for this. It's a bit surprising, and we should document it. (I also added `docs-known-limitation` to #144443.)
___
*[`pkg/sql/alter_primary_key.go` line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144601#-OO16pNH-uEtTShwi8uP:-OO16pNH-uEtTShwi8uQ:b-sz0pbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce6846aa719832806affc3a5dc1ce5cfa1d9b237/pkg/sql/alter_primary_key.go#L58)):*
> ```Go
> ) error {
> 	// Check if sql_safe_updates is enabled and the table has vector indexes
> 	if p.EvalContext().SessionData().SafeUpdates {
> ```

I like this message when `sql_safe_updates` is enabled. Honestly, it might be nice to emit the same message as a notice even when `sql_safe_updates` is disabled.

The way to do that is something like:

```go
		p.BufferClientNotice(
			ctx,
			pgnotice.Newf("ALTER PRIMARY KEY on a table with vector indexes will disable writes to the table while the index is being rebuilt"),
		)
```

(This is only a suggestion. Also could be in a follow-up PR.)


<!-- Sent from Reviewable.io -->

mw5h committed:
schemachanger: when taking a vector index offline, notify client

A previous patch disabled mutations on vector indexes if
sql_safe_updates was enabled. This patch send the associated error
message as a notice even if sql_safe_updates is disabled, so there are
no surprises.

Informs: #144443
Release note (sql change): CREATE VECTOR INDEX and ALTER PRIMARY KEY
will now send a notice that vector indexes will be offline during the
change operation when sql_safe_updates is disabled.
